demo/image_demo.py

```diff
@@ -1,94 +1,86 @@
+    parser.add_argument(
+        '--show',
+        action='store_true',
+        help='Display the image in a popup window.')
+    parser.add_argument(
+        '--no-save-img',
+        action='store_true',
+        help='Do not save detection vis results')
+    parser.add_argument(
+        '--no-save-pred',
+        action='store_true',
+        help='Do not save detection json results')
+    parser.add_argument(
+        '--print-result',
+        action='store_true',
+        help='Whether to print the results.')
     parser.add_argument(
         '--palette',
         default='none',
         choices=['coco', 'voc', 'citys', 'random', 'none'],
         help='Color palette used for visualization')
-    parser.add_argument(
-        '--score-thr', type=float, default=0.3, help='bbox score threshold')
-    parser.add_argument(
-        '--async-test',
-        action='store_true',
-        help='whether to set async options for async inference.')
-    args = parser.parse_args()
-    return args
-
 
-def main(args):
-    # TODO: Support inference of image directory.
-    # build the model from a config file and a checkpoint file
-    model = init_detector(
-        args.config, args.checkpoint, palette=args.palette, device=args.device)
+    call_args = vars(parser.parse_args())
 
-    # init visualizer
-    visualizer = VISUALIZERS.build(model.cfg.visualizer)
-    # the dataset_meta is loaded from the checkpoint and
-    # then pass to the model in init_detector
-    visualizer.dataset_meta = model.dataset_meta
+    if call_args['no_save_img'] and call_args['no_save_pred']:
+        call_args['out-dir'] = ''
 
-    # test a single image
-    result = inference_detector(model, args.img)
+    if call_args['model'].endswith('.pth'):
+        print_log('The model is a weight file, automatically '
+                  'assign the model to --weights')
+        call_args['weights'] = call_args['model']
+        call_args['model'] = None
 
-    # show the results
-    img = mmcv.imread(args.img)
-    img = mmcv.imconvert(img, 'bgr', 'rgb')
-    visualizer.add_datasample(
-        'result',
-        img,
-        data_sample=result,
-        draw_gt=False,
-        show=args.out_file is None,
-        wait_time=0,
-        out_file=args.out_file,
-        pred_score_thr=args.score_thr)
+    init_kws = ['model', 'weights', 'device', 'palette']
+    init_args = {}
+    for init_kw in init_kws:
+        init_args[init_kw] = call_args.pop(init_kw)
 
+    return init_args, call_args
 
-async def async_main(args):
-    # build the model from a config file and a checkpoint file
-    model = init_detector(args.config, args.checkpoint, device=args.device)
 
-    # init visualizer
-    visualizer = VISUALIZERS.build(model.cfg.visualizer)
-    visualizer.dataset_meta = model.dataset_meta
+def main():
+    init_args, call_args = parse_args()
+    inferencer = DetInferencer(**init_args)
+    inferencer(**call_args)
```

Comment:
I still insist that it is better to put the loop that traverses the image file outside the inferencer. This makes it easier for users to debug and understand the code. Encapsulating everything inside this class is confusing.

Reply:
TODO